### ISSUE TITLE
Allow searching for content on Admin bar "menu search" field

### ISSUE THREAD

**Issue by sutibun** [backdrop/backdrop-issues#1258]
This isn't particularly urgent but would be really useful and meet user expectations. When I first saw the admin bar search form...

![search](https://cloud.githubusercontent.com/assets/14132746/10119264/b1595508-6456-11e5-9d52-e16500e3c021.png)

... I immediately thought I could type something like "cookies" and Bd would search, pull up and then link to the proper content. (A search form for your written content basically)

I was like, "cool!".

Eventually figured out that wasn't the case. Now, I do like the menu search feature but **feel like this is a quick and small letdown to new users coming to Backdrop**. Even with the text "Menu search", many new users will still immediately think like I do. We are beginning the first foot with disappointment.
- I doubt new users will really get what "Menu search" means at first glance. By the time they do, they will be silently disappointed like me. "Why can't it just search for my content??" Friction enters.
- Finding your content with the search from the admin bar is much faster. Sometimes I know the post title and don't need to go all the way to "admin>content>find content". Boost productivity basically.

Anyway, here is a quick mockup of how it can look

![menusearch](https://cloud.githubusercontent.com/assets/14132746/10119272/fbe0a25c-6456-11e5-9ef9-6d8ec719a70b.png)

Some suggestions
- limit the content search to the top 3 found results that match the query.
  - uh... probably move the content search results above the menu search since results are limited
- provide a link to "admin>content>find content" so users can browse and explore if search fails
- also provide a link to the full search results page
- provide options
  - add an option to disable searching content for the users who have no need for them.
  - add an option to disable menu search too
  - provide option to set how many results are found for content search

**Comment by klonos** [comment]
> ...feel like this is a quick and small letdown to new users coming to Backdrop

Just a reminder that this search field is on the admin bar and that bar is (by default) only available to site builders and admins - not the rest of the site visitors and content editors. Site admins are only a handful (if not only one) and I believe that they are rarely working (end thus searching) on content.

**Comment by sutibun** [comment]
oh, I wasn't referring to visitors. I was referring to new people who use the backend of Backdrop like editors and writers. In some cases, the admin users is everything (builder, editor, writer, etc) like me in some of my sites.

**Comment by rszrama** [comment]
fwiw, I almost tried to use the search bar for a content search as well, administrator yae though I be. :dancer:

**Comment by sutibun** [comment]
\*wipes forehead\* Yay, the monkey is not the only one

**Comment by klonos** [comment]
...but it says "Menu search". It's there in b&w. :stuck_out_tongue_winking_eye:

**Comment by rszrama** [comment]
Yep. That's why I stopped before trying. But it's still jarring. I don't think it's that odd that our first assumption would be a search bar searches the site. "Menu search" just isn't a standard user interface element, so it takes a second to register first "Ahh, this is restricted to items in this menu," and then "Oh, and just the items themselves, not any content contained on them."

(If you wanna see a cool result set, type list in there... ; )

**Comment by sutibun** [comment]
> ...but it says "Menu search". It's there in b&w. :stuck_out_tongue_winking_eye:

Monkeys only see in colors. Not b&w.

> If you wanna see a cool result set, type list in there..

You broke my site...

**Comment by klonos** [comment]
> Monkeys only see in colors. Not b&w. That's my excuse and I'm sticking to it.

:smile: 

>> If you wanna see a cool result set, type list in there..

> You broke my site...

Yep, I know... #525, #1002

**Comment by dboulet** [comment]
What might be needed here is a simple modification to the way that the menu search is presented. You can’t blame users for thinking that a simple search input field would be used for a site search. It’s not clear enough that the search is limited to menu items—even the meaning of the words “Menu search” are not immediately obvious.

In OS X, for example, the menu search is displayed under the **Help** menu. This makes it a little clearer that the search field is not intended for searching site content, but for directing users who need assistance.

![screen shot 2015-09-28 at 1 44 47 pm](https://cloud.githubusercontent.com/assets/208615/10145359/38543eaa-65e8-11e5-886f-ed8388c693b0.png)

**Comment by sutibun** [comment]
> You can’t blame users for **thinking that a simple search input field would be used for a site search**. It’s not clear enough that the search is limited to menu items—**even the meaning of the words “Menu search” are not immediately obvious.**

Totally agree.

However, I actually like how I can find menu items from that search form **quickly and easily**. I find it helpful since the way the admin menus are organized is confusing and not intuitive. Hiding the search form under a menu makes it harder when I want to find a menu item quick. i.e. having to find a search form just to do a search and find a menu item defeats the purpose.

I'm **not** really looking for a way clarify that search form so users know it's only for menu items.

In fact, the opposite.

> That's why I stopped before trying. **But it's still jarring. I don't think it's that odd that our first assumption would be a search bar searches the site.** "Menu search" just isn't a standard user interface element

I want to **meet their expectations** by adding the ability to search for content on that **same** search form.

I don't have a mac but something like spotlight search?

![spotlight](https://cloud.githubusercontent.com/assets/14132746/10171417/a9676b32-669f-11e5-9f04-4aa883fff337.png)

Basically, if the user is a 

- writer/editor, the search form will only list content entries.
- admin, the search form will list both content and menu entries

The search form will adjust depending on the permission a user has. The higher the permissions, the more the user will see. We should add options to disable some of them. Some admins may not care for menu search and only content entries. Vice versa.

Basically, **without adding a new UI element onto the admin bar, we've added new useful features and improved the UX at the same time**

**Comment by docwilmot** [comment]
I was suggesting a related feature recently: ["context aware help"](https://github.com/backdrop/backdrop-issues/issues/1249#issuecomment-142948366), where the Admin bar would provide links to appropriate help automatically based on the context.

**Comment by sutibun** [comment]
@docwilmot You should open a ticket where we can argue... er... discuss. ^\_^ I think there's some value there worth exploring.

**Comment by klonos** [comment]
Yes, you do have a point that users might miss the "menu search" placeholder text that is there on the search field (or even misinterpret its functionality) and yes, adding content and help topics search would be adding [UX] value. I too like the way that Mac Spotlight and Windows search work returning search results on various menus/files/emails etc, but... a couple of thoughts:

The search that now happens on that menu as it is (searching for menu items), is possible because these menu items are already loaded on the page. In fact it is not a search field rather than an "instant" filter (find-as-you-type) field. So, if we were to add search capabilities on content and documentation/help, I'm guessing that we'd either need to load those results through AJAX (still not instant) or have it so that users type a search string in there and then need to hit enter and have the results be printed in a search results list (page load). None of these would be as fast as the JS that is there doing some "magic" and filtering matching menu items because for content/help it would need to query the database. Can we afford to have the database queried on every keystroke? Would we have to load all content on memory/cache and search that instead? Not sure how that would work. Those with more insight please shed some light.

**Comment by sutibun** [comment]
> Can we afford to have the database queried on every keystroke? Would we have to load all content on memory/cache and search that instead?

Er... rhe technical/implementation details are a lil' beyond me... the most advanced technology the monkey knows is two stick tied together.

> Also, the way the search/filtering results work at the moment is that when you hover over them, the menu is expanded down to the respective menu item being hovered on ("teaching" the user how to get there). What would the equivalent behavior for content/help topics be?

Well, the search form is meant to provide easy access to content entries you already know you want to view/edit but want to bypass the "Find content" page. Basically, for quick and easy access. No need for an equivalent behavior.

PS: I don't know about help topics... that's the doc's idea.

**Comment by docwilmot** [comment]
How about the first two items in the search results be fixed as so:

```
[__search term in search bar__] <-- this is the search field
- Click to search site content       <-- first item in dropdown
- Click to search online help
- Menu search result 1                <-- menu search items as usual
- Menu search result 2
```

**Comment by sutibun** [comment]
Oh, I didn't know you meant actually adding the help topic entries in the search. I liked your original idea:

> So you're on a user-related page and you hover over the help button and get links related to account stuff.

**Comment by docwilmot** [comment]
This is a new idea. Just throwing them out there to see if one could work..

**Comment by sutibun** [comment]
Ah kk. I like your technical mockup btw ^\_^ Took a second but I got it.

**Comment by klonos** [comment]
@docwilmot :+1: for the "Click to search..." links added at the top of the result list. Perhaps we could also denote their difference (say italics or slightly different text color) from the rest of the search results (menu items matches).

I like this solution because it seems to answer all my concerns about having to query the db for content results etc.

**Comment by sutibun** [comment]
[__search term in search bar__] <-- this is the search field
    - Click to search site content       <-- first item in dropdown
    - Click to search online help
    - Menu search result 1                <-- menu search items as usual
    - Menu search result 2

Honestly, the link to the search results is a lil' of a letdown as opposed to showing the content listing below the search form BUT, practically speaking, this is a HUGE improvement and I welcome it with open arms.

**Comment by yorkshire-pudding** [comment]
I don't think this is a good idea to have in core for the following reasons:

- Sites with lots of content will have a performance impact when you start to search - [see the comment ](https://github.com/backdrop/backdrop-issues/issues/1258#issuecomment-144200788 )by @klonos 
- For those looking for menu items the results could be very cluttered if the term matches something in content.

I've added the "contrib candidate" label.
